gobblin-modules/gobblin-helix/src/main/java/gobblin/metastore/ZkStateStore.java

```diff
@@ -0,0 +1,279 @@
+    List<T> states = Lists.newArrayList();
+
+    deserialize(data, states, stateId);
+
+    if (states.isEmpty()) {
+      return null;
+    } else {
+      return states.get(0);
+    }
+  }
+
+  /**
+   * Retrieve states from the state store based on the store name and a filtering predicate
+   * @param storeName The store name enclosing the state files
+   * @param predicate The predicate for state file filtering
+   * @return list of states matching matching the predicate
+   * @throws IOException
+   */
+  protected List<T> getAll(String storeName, Predicate<String> predicate) throws IOException {
+    List<T> states = Lists.newArrayList();
+    String path = formPath(storeName);
+    byte[] data;
+
+    List<String> children = propStore.getChildNames(path, 0);
+
+    if (children == null) {
+      return Collections.emptyList();
+    }
+
+    for (String c : children) {
+      if (predicate.apply(c)) {
+        data = propStore.get(path + "/" + c, null, 0);
+        deserialize(data, states);
+      }
+    }
+
+    return states;
+  }
+
+  @Override
+  public List<T> getAll(String storeName, String tableName) throws IOException {
+    List<T> states = Lists.newArrayList();
+    String path = formPath(storeName, tableName);
+    byte[] data = propStore.get(path, null, 0);
+
+    deserialize(data, states);
+
+    return states;
+  }
+
+  @Override
+  public List<T> getAll(String storeName) throws IOException {
+    return getAll(storeName, Predicates.<String>alwaysTrue());
+  }
+
+  @Override
+  public void createAlias(String storeName, String original, String alias) throws IOException {
+    String pathOriginal = formPath(storeName, original);
+    String pathAlias = formPath(storeName, alias);
+    byte[] data;
+
+    if (!propStore.exists(pathOriginal, 0)) {
+      throw new IOException(String.format("State does not exist for table %s", original));
+    }
+
+    data = propStore.get(pathOriginal, null, 0);
+
+    if (!exists(storeName, alias) && !create(storeName, alias)) {
+      throw new IOException("Failed to create a state file for table " + alias);
+    }
+
+    propStore.set(pathAlias, data, AccessOption.PERSISTENT);
+  }
+
+  @Override
+  public void delete(String storeName, String tableName) throws IOException {
+    propStore.remove(formPath(storeName, tableName), 0);
+  }
+
+  @Override
+  public void delete(String storeName) throws IOException {
+    propStore.remove(formPath(storeName), 0);
+  }
+
+  /**
+   * Deserialize data into a list of {@link State}s.
+   * @param data byte array
+   * @param states output list of states
+   * @param stateId optional key filter. Set to null for no filtering.
+   * @throws IOException
+   */
+  private void deserialize(byte[] data, List<T> states, String stateId) throws IOException {
+    if (data != null) {
+      ByteArrayInputStream bais = new ByteArrayInputStream(data);
+      InputStream is = compressedValues ? new GZIPInputStream(bais) : bais;
```

Comment:
I'd consider storing of the compressedValues flag in ZK. Thus we can avoid error scenarios where previosly compressed values are being read with compressedValues == false and vice versa.